### ISSUE TITLE
Feature request : Provision to retrain the network

### ISSUE THREAD

**Issue by saikishor** [tensorflow/models#2510]
It will be good to have a provision to retrain the Tensorflow API models and to add new classes to the pre-trained models.

**Comment by tatatodd** [comment]
@saikishor Can you explain what exactly you mean?  Perhaps give a concrete example of what you'd like to see.

**Comment by saikishor** [comment]
I mean to make a provision to add more classes to current pre-trained net(90-class pretrained net + additional 2 classes = 92 class model), and to retrain the network with few classes to improve the precision of those classes(for example, only to retrain 5 classes of 90 class net and as a result get a 90 class net but with more precision to the trained 5 classes.

**Comment by tatatodd** [comment]
@saikishor Is there a specific model you have in mind?  Many of the models have explicit instructions about training from scratch and fine-tuning.

Also can you be more explicit about what you mean by "provision".  E.g. are you looking for more documentation?  More examples?  More configuration flags?

**Comment by saikishor** [comment]
@tatatodd Yes I am looking for more documentation on fine-tuning and retraining with one example. I feel this will help many young researchers. 
I am mentioning about these [Tensorflow detection model zoo](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md) models

**Comment by tatatodd** [comment]
@saikishor I think ObjectDetection actually has quite a bit of good documentation already.

There is general documentation on object detection here:
https://github.com/tensorflow/models/tree/master/research/object_detection

And there is documentation on fine-tuning object detection, starting from COCO, and refining for pets, here:
https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/running_pets.md

More (good) documentation is always helpful, but it's not clear to me what specific areas are lacking.

**Comment by saikishor** [comment]
Example: I am also referring to this one - https://github.com/tensorflow/models/issues/1809.
I just wanted to know a way to add more classes to existing model. That's what I am looking for.

As per the above link, I can see that there is no way to do that as of now. So, I am just saying providing that would be more helpful.

**Comment by tatatodd** [comment]
I see.

As mentioned on that bug, you can't "add classes to an existing model"; but you can finetune a new model starting from one of the checkpoints.  There are quite a few documents in the links I provided above describing this.

Closing this out.

**Comment by saikishor** [comment]
Sure thank you!!

**Comment by lucifer6666** [comment]
Is there a way to add extra data to pre trained machine.
I mean that suppose I have made a facial recognition system and added the image data of 100 people, 
now I want to add one more person data to make it 101 persons,
Is there a way to only train the data of 1 person and add it to the pre trained machine.

**Comment by SriramganeshMarimuthu** [comment]
@lucifer6666 As per your question, Did you find any solution for that to train your method? if you get the idea or solution please share to me.

**Comment by dwSun** [comment]
This may help.
1606.09282-Learning without Forgetting

**Comment by Rahul-Aanand** [comment]
@saikishor Any update on retraining a pre-trained model to detect objects of new classes in addition to preserving object detection for previously trained classes (detecting 90+1 classes)?

**Comment by saikishor** [comment]
@Rahul-Aanand sorry, unfortunately no. I was not able to find a solution for it. 

If you are providing the whole dataset for training then it should work. Atleast i have done it, it worked for me.  It's not only the thing of labels file, but also the training dataset
